openwisp_controller/config/tasks.py

```diff
@@ -104,3 +104,57 @@ def trigger_vpn_server_endpoint(endpoint, auth_token, vpn_id):
             f'Response status code: {response.status_code}, '
             f'VPN Server UUID: {vpn_id}',
         )
+
+
+@shared_task(base=OpenwispCeleryTask)
+def change_devices_templates(instance_id, model_name, **kwargs):
+    Device = load_model('config', 'Device')
+    DeviceGroup = load_model('config', 'DeviceGroup')
+    Template = load_model('config', 'Template')
+    Config = load_model('config', 'Config')
+
+    if model_name == Device._meta.model_name:
+        instance_ids = instance_id
+        old_group_ids = kwargs.get('old_group_id')
+        group_id = kwargs.get('group_id')
+        if type(instance_id) is not list:
+            instance_ids = [instance_id]
+            old_group_ids = [old_group_ids]
+        for instance_id, old_group_id in zip(instance_ids, old_group_ids):
+            device = Device.objects.get(pk=instance_id)
+            if not hasattr(device, 'config'):
+                return
+            group_templates = Template.objects.none()
+            if group_id:
+                group = DeviceGroup.objects.get(pk=group_id)
+                group_templates = group.templates.all()
+            old_group_templates = Template.objects.none()
+            if old_group_id:
+                old_group = DeviceGroup.objects.get(pk=old_group_id)
+                old_group_templates = old_group.templates.all()
+            device.config.manage_group_templates(group_templates, old_group_templates)
+
+    elif model_name == DeviceGroup._meta.model_name:
+        device_group = DeviceGroup.objects.get(id=instance_id)
+        templates = Template.objects.filter(pk__in=kwargs.get('templates'))
+        old_templates = Template.objects.filter(pk__in=kwargs.get('old_templates'))
+        for device in device_group.device_set.all():
+            if not hasattr(device, 'config'):
+                continue
+            device.config.manage_group_templates(templates, old_templates)
+
+    elif model_name == Config._meta.model_name:
+        config = Config.objects.get(pk=instance_id)
+        device_group = config.device.group
+        if not device_group:
+            return
+        created = kwargs.get('created')
+        if created:
+            templates = device_group.templates.all()
+            old_templates = Template.objects.none()
+        else:
+            templates = device_group.templates.filter(backend=kwargs.get('backend'))
+            old_templates = device_group.templates.filter(
+                backend=kwargs.get('old_backend')
+            )
+        config.manage_group_templates(templates, old_templates, not created)
```

Comment:
It bothers me a bit that we have so much logic in a celery task.
If you look at all the other tasks of this project, you will see that most of the tasks just call class methods on the relevant model, which follows the philosophy of having most of the core logic there, which is better for many reasons, including that if needed we can call a specific part of this function independently from the other parts.